### PR TITLE
Decoder stateful

### PR DESCRIPTION
**Pull request by jlibovicky**
This PR make autoregressive instance of `TemporalStateful` such we can stack a sequence labeler on top of it.

### CODE REVIEW

File: neuralmonkey/decoders/autoregressive.py [hunk 1]
```diff
@@ -418,3 +419,19 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
 
         return fd
+
+    @tensor
+    def temporal_states(self) -> tf.Tensor:
+        return tf.cond(
+            self.train_mode,
+            lambda: tf.transpose(self.train_output_states, [1, 0, 2])[:, :-2],
+            lambda: tf.transpose(
+                self.runtime_output_states, [1, 0, 2])[:, :-2])
```
Comment (varisd) [inline neuralmonkey/decoders/autoregressive.py]: Proc tady zahazujes ty posledni dva prvky? Mozna by to chtelo komentar

File: neuralmonkey/decoders/autoregressive.py [hunk 2]
```diff
@@ -418,3 +419,19 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
             fd[self.train_mask] = weights
 
         return fd
+
+    @tensor
+    def temporal_states(self) -> tf.Tensor:
+        return tf.cond(
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/autoregressive.py]: přesvědč mě prosím, že to chceme mít závislé na dekodéřím train modu.. Neni reálná situace, kdy chceme labeler trénovat na runtime stavech, případně běhat labeler s trénovacím módem dekodéru? (To druhý asi ne; nicméně pokud mam zafixovanej dekodér a trénuju si labeler, tak `self.train_mode` bude nejspíš true, takže se mi muj labeler nakazí exposure biasem od dekodéru, ne?)

File: neuralmonkey/decoders/sequence_labeler.py [hunk 3]
```diff
@@ -89,17 +89,19 @@ def logits(self) -> tf.Tensor:
         multiplication_3d = tf.squeeze(multiplication, squeeze_dims=[2])
 
         biases_3d = tf.expand_dims(tf.expand_dims(self.decoding_b, 0), 0)
+        logits = multiplication_3d + biases_3d
 
-        embedded_inputs = tf.expand_dims(
-            self.encoder.input_sequence.temporal_states, 2)
-        dweights_4d = tf.expand_dims(
-            tf.expand_dims(self.decoding_residual_w, 0), 0)
+        if hasattr(self.encoder, "input_sequence"):
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: Co když je `self.encoder` dekodér - nezajímají nás embeddingy slov, co lezou na vstup dekodéru? Asi by to bylo už na větší refactor..

Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Ja bych spis cekal, ze te spis zajimaji skryte stavy (resp. to, na zaklade ceho potom predikujes vystupy).

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: Když ten labeler pracuje nad enkodérem, tak kouká jak na stavy, tak na vstupy. A tadyten if je tu proto, že u dekodéru to padalo, protože nemá vstupy. Ale on je taky svym způsobem má. Plus je samozřejmě ošklivý `if` přes `hasattr`

**Comment by cifkao** [comment]
Jen nápad: Co přidat třídu `DecoderSequence` (nebo by to mohla být funkce `get_decoder_states`, která by vracela nějaký `TemporalStateful`), které by se řeklo, co přesně se z dekodéru chce (stavy, vstupy, logity...) a v jakém režimu?

(Nejlepší by bylo mít to jako metodu dekodéru, ale ta nejde zavolat z konfiguráku.)

**Comment by jindrahelcl** [comment]
Já bych přidal nějakou kouzelnou věc, která by z tenzorů v čase (resp prostoru) udělala temporal (resp spatial) stateful objekty.
 fieldům se z konfiguráku přistupovat dá.

Takže bychom měli v dekodéru normálně:

```python3
@tensor
def runtime_logits(self) -> tf.Tensor:
    #...
```

pak někde v nějakym helper modulu něco jako

```python3
def make_temporal_stateful(tensor: tf.Tensor) -> TemporalStateful:
    #...
```

a v konfigu:

```ini
[decoder_logits]
class=make_temporal_stateful
tensor=<decoder.runtime_logits>
```

No a nebo by se ten  `make_stateful` volal neviditelně z konstruktoru toho objektu, kterej by žral `Union[tf.Tensor, TemporalStateful]` a pokud by to byl tensor, tak by to tu `make` funkci zavolalo samo.

**Comment by cifkao** [comment]
@jindrahelcl `TemporalStateful` potřebuje ještě masku (i když ta asi defaultně může být `None` nebo jedničková). A měl by to asi navíc být `ModelPart` se závislostí na tom model partu, odkud je ten tenzor.

Musel by se přidat nějaký `TemporalStatefulAdapter` nebo `DefaultTemporalStateful`, který by vyžadoval `temporal_states`, `temporal_mask` a `dependencies`. Dekodér by vyrobil instance téhle třídy (s `dependencies=[self]`) a vystavil jako properties, na které by se dalo odkazovat z konfiguráku. Ale to je trochu ošklivé.

Navíc možná chceme mít možnost, aby to bylo závislé na tom train modu.

**Comment by jlibovicky** [comment]
Kouzelná věc mi přijde zbytečně složitá, pokud nemáme jinej usecase, než tohle.

File: neuralmonkey/decoders/sequence_labeler.py (hunk 3 above)
Comment (jlibovicky) [reply neuralmonkey/decoders/sequence_labeler.py]: Mohly by nás zajímat, ale znamenalo by to docela dost překopat autoregressive decoder a do toho se mi nechce pouštět. Navíc by to potom nebyla input sequence a bylo potřeba nějak sjednotit interface věcí, co mají embeddingy.

File: neuralmonkey/decoders/autoregressive.py (hunk 2 above)
Comment (jlibovicky) [reply neuralmonkey/decoders/autoregressive.py]: Představ si, že chceš trénovat multitask překlad a POS tagging na výstupu. Máš POS tagy pro trénovací data. Pak přece nemůžeš nechat dekodér si generovat, co chce, aby labeler nechytil exposure bias. Musíš naopak zařídit, aby ten výstup dekodéru odpovídal tomu, co je v těch zlatých datech, protože jinak nemáš záruku, že stav enkodéru odpovídá tomu, co máš ve zlatých datech.

Kdybys chtěl trénovat ty tagy na runtime stavech, tak bys musel dokázat zjišťovat ty správný tagy dynamicky, podle toho, co se zrovna vydekódovalo, a to není možné.

**Comment by jlibovicky** [comment]
mypy mi říká:
```
neuralmonkey/decoders/transformer.py:110: error: Property "dimension" defined in "TransformerDecoder" is read-only
```
Nevíte někdo, co to je?

**Comment by jindrahelcl** [comment]
`dimension` je poděděná `@property` z `TemporalStateful`, takže je read-only,

**Comment by jindrahelcl** [comment]
Kdežto tady se `dimension` značí dimenze modelu, jak o ní mluví Vaswani et al

**Comment by jlibovicky** [comment]
Aha, tak přejmenujeme to u transormera na `model_dimension`?

**Comment by jindrahelcl** [comment]
Můžeme

Dne 5. 3. 2018 12:16 odp. napsal uživatel "Jindřich Libovický" <
notifications@github.com>:

File: neuralmonkey/decoders/autoregressive.py (hunk 2 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: A jaký slovo tim taggerem labeluješ? Předchozí vydekódovaný nebo zrovna vydekódovaný? Protože jestli zrovna vydekódovaný, tak to je taky něco jinýho než reference.

File: neuralmonkey/decoders/sequence_labeler.py [hunk 4]
```diff
@@ -1,23 +1,33 @@
...
 class SequenceLabeler(ModelPart):
-    """Classifier assing a label to each encoder's state."""
+    """Classifier assigning a label to each input state.
+
+    If the labeler output has an input sequence with embeddings, these are used
+    as additional input to the labeler.
+
+    Note that when the labeler is stacked on an autoregressive decoder, it
+    labels the symbol that is currently generated by the decoder, i.e., the
+    decoder's state has not yet been updated by putting the decoded symbol on
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: ``s/'s//``

File: neuralmonkey/decoders/sequence_labeler.py [hunk 5]
```diff
@@ -1,23 +1,33 @@
...
+    its input. The label is thus the label of a symbol is generated, not the
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: nesrozumitelná věta

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: (ta co začíná The label)

**Comment by jlibovicky** [comment]
Tak, když  jsem přidal pořádnej tests, tak to padá na tom, že za runtimu není nafeedovaná `train_inputs` v dekodéru. Ale upřímně nevím, kde se tam berou, protože mi `tf.cond` v autoregresivním dekodéru mi přijde v pořádku.

**Comment by jindrahelcl** [comment]
Zkus TF 1.5 on ten cond to tam může propouštět i když je runtime.. ale kvůli tomu tam jsou ty lambdy, ne?

**Comment by jindrahelcl** [comment]
ok tak TF 1.6 :-)

**Comment by jindrahelcl** [comment]
@jlibovicky tohle je taková kravina, že by se dala rebasnout a mergnout, ne?

**Comment by jlibovicky** [comment]
Zkus to, ale já mylslim, že to na něčem padalo.

File: neuralmonkey/decoders/autoregressive.py [hunk 6]
```diff
@@ -475,3 +476,21 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
             fd[self.train_mask] = weights
 
         return fd
+
+    @tensor
+    def temporal_states(self) -> tf.Tensor:
+        # strip the last symbol which is </s>
+        return tf.cond(
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/autoregressive.py]: Tohle takhle nefunguje. `tf.cond` potřebuje všechny placeholdery, i když se nakonec vyhodnotí opačně.

Comment (jlibovicky) [reply neuralmonkey/decoders/autoregressive.py]: Jo, to už jsem jednou věděl. Myslím, že se to mělo spravit v nějaké další verzi TF. Jak jsme na tom s kompatibilitou s novými verzemi?

File: neuralmonkey/decoders/sequence_labeler.py [hunk 7]
```diff
@@ -102,13 +114,20 @@ def decoded(self) -> tf.Tensor:
...
-        loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
-            labels=self.train_targets, logits=self.logits)
-
-        # loss is now of shape [batch, time]. Need to mask it now by
-        # element-wise multiplication with weights placeholder
-        weighted_loss = loss * self.train_weights
-        return tf.reduce_sum(weighted_loss)
+        min_time = tf.minimum(tf.shape(self.train_targets)[1],
+                              tf.shape(self.logits)[1])
+
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_labeler.py]: Skutecne tu chceme vybirat minimum behem trenovani. Nehrozi treba, ze se sit nauci spravne generovat (dekoderovou vrstvou) napr. pouze sekvence delky 1, ktere ale bude spravne labelovat.

Nemely by se logits "paddovat" na train_targets length, nebo nejakym jinym zpusobem penalizovat kratsi sekvence?

File: neuralmonkey/decoders/sequence_labeler.py [hunk 8]
```diff
@@ -102,13 +114,20 @@ def decoded(self) -> tf.Tensor:
...
+        # symbol (or for some reason emits more symbol than we have in the
+        # ground truth labels), we trim the sequences to the length of a
+        # shorter one.
+
+        # pylint: disable=unsubscriptable-object
+        return tf.contrib.seq2seq.sequence_loss(
+            logits=self.logits[:, :min_time],
+            targets=self.train_targets[:, :min_time],
+            weights=self.input_sequence.temporal_mask[:, :min_time])
+        # pylint: enable=unsubscriptable-object
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_labeler.py]: Je skutecne v poradku tady nahradit sumu pres logity meanem? Minimalne z hlediska zpetne kompatibility (trenovacich hyperparametru) to uplne koser nebude.

File: neuralmonkey/decoders/transformer.py [hunk 9]
```diff
@@ -126,10 +126,15 @@ def __init__(self,
...
         self.encoder_mask = get_attention_mask(self.encoder)
-        self.dimension = (
-            self.encoder_states.get_shape()[2].value)  # type: ignore
 
-        if self.embedding_size != self.dimension:
+        # This assertion (and the "int" type declaration below) here is because
+        # of mypy not being able to handle the tf.Tensor type.
+        assert self.encoder_states is not None
+
+        self.model_dimension = (
```
Comment (varisd) [inline neuralmonkey/decoders/transformer.py]: Nebylo by lepsi misto prejmenovavani radej presunout radky 130-136 do overridnute property dimension? Takhle vznika zmatek v kodu
